### ISSUE TITLE
Word/Excel/PowerPoint saves the file immediately after opening it

### ISSUE THREAD

**Issue by Cyb3r-Monk** [cmu-sei/GHOSTS#200]
Would it be possible to make Word/Excel/PowerPoint handlers wait for some predefined duration and save the file? They currently open a document and immediately save it.

**Comment by Cyb3r-Monk** [comment]
After saving the document, it closes itself. I modified the handler to accomplish what I want to do. I added a `stay-open` handler argument. After opening/creating a document, it waits for a while according to the `stay-open` parameter. Then saves the document and closes the application. Below is what I did in the Word handler (other are the same and I also fixed the save-array folder issue)

```C#
// Copyright 2017 Carnegie Mellon University. All Rights Reserved. See LICENSE.md file for terms.

using Ghosts.Client.Infrastructure;
using Ghosts.Domain;
using NetOffice.WordApi.Enums;
using System;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Threading;
using Ghosts.Domain.Code;
using Ghosts.Domain.Code.Helpers;
using Newtonsoft.Json;
using Word = NetOffice.WordApi;
using VB = Microsoft.VisualBasic;

namespace Ghosts.Client.Handlers;

public class WordHandler : BaseHandler
{
    public int JitterFactor { get; set; } = 0;
    public int StayOpen;
    public WordHandler(Timeline timeline, TimelineHandler handler)
    {
        try
        {
            base.Init(handler);
            Log.Trace("Launching Word handler");

            if (handler.Loop && handler.ScheduleType != TimelineHandler.TimelineScheduleType.Cron)
            {
                Log.Trace("Word loop");
                while (true)
                {
                    if (timeline != null)
                    {
                        System.Collections.Generic.List<int> processIds = ProcessManager.GetPids(ProcessManager.ProcessNames.Word).ToList();
                        if (processIds.Count > 2 && processIds.Count > timeline.TimeLineHandlers.Count(o => o.HandlerType == HandlerType.Word))
                        {
                            continue;
                        }
                    }

                    ExecuteEvents(timeline, handler);
                }
            }
            else
            {
                Log.Trace("Word single run");
                KillApp();
                ExecuteEvents(timeline, handler);
                KillApp();
            }
        }
        catch (ThreadAbortException)
        {
            KillApp();
            Log.Trace("Thread aborted, Word closing...");
        }
        catch (Exception e)
        {
            Log.Error(e);
            KillApp();
        }
    }

    private static void KillApp()
    {
        ProcessManager.KillProcessAndChildrenByName(ProcessManager.ProcessNames.Word);
    }

    private void ExecuteEvents(Timeline timeline, TimelineHandler handler)
    {
        try
        {
            try
            {
                ParseRandomHandlerArgs(handler);
            }

            catch (Exception e)
            {
                Log.Trace("Word:: Error parsing handler args");
                Log.Error(e);

            }

            foreach (var timelineEvent in handler.TimeLineEvents)
            {
                try
                {
                    Log.Trace($"Word event - {timelineEvent}");
                    Infrastructure.WorkingHours.Is(handler);

                    if (timelineEvent.DelayBefore > 0)
                    {
                        // Thread.Sleep(timelineEvent.DelayBefore);
                        Thread.Sleep(Jitter.JitterFactorDelay(timelineEvent.DelayBefore, JitterFactor));
                    }

                    if (timeline != null)
                    {
                        var processIds = ProcessManager.GetPids(ProcessManager.ProcessNames.Word).Count();
                        if (processIds > timeline.TimeLineHandlers.Count(o => o.HandlerType == HandlerType.Word))
                        {
                            return;
                        }
                    }

                    // start word and turn off msg boxes
                    using (var wordApplication = new Word.Application
                           {
                               DisplayAlerts = WdAlertLevel.wdAlertsNone,
                               Visible = true
                           })
                    {
                        Word.Document document = null;
                        if (OfficeHelpers.ShouldOpenExisting(handler))
                        {
                            document = wordApplication.Documents.Open(FileListing.GetRandomFile(handler.HandlerType));
                            Log.Trace($"{handler.HandlerType} opening existing file: {document.FullName}");
                        }
                        if (document == null)
                        {
                            document = wordApplication.Documents.Add();
                            Log.Trace($"{handler.HandlerType} adding new...");
                        }

                        try
                        {
                            wordApplication.WindowState = WdWindowState.wdWindowStateMinimize;
                            foreach (var item in wordApplication.Documents)
                            {
                                item.Windows[1].WindowState = WdWindowState.wdWindowStateMinimize;
                            }
                        }
                        catch (Exception e)
                        {
                            Log.Trace($"Could not minimize: {e}");
                        }

                        var writeSleep = Jitter.JitterFactorDelay(30000, 40);

                        // insert some text
                        var list = RandomText.GetDictionary.GetDictionaryList();
                        using (var rt = new RandomText(list))
                        {
                            rt.AddContentParagraphs(1, 50);
                            wordApplication.Selection.TypeText(rt.Content);
                            Thread.Sleep(writeSleep);
                        }

                        wordApplication.Selection.HomeKey(WdUnits.wdLine, WdMovementType.wdExtend);
                        wordApplication.Selection.Font.Color = GetWdColor(StylingExtensions.GetRandomColor());
                        wordApplication.Selection.Font.Bold = 1;
                        wordApplication.Selection.Font.Size = 12;

                        Log.Trace($"Staying open for {StayOpen} ms with {JitterFactor}% jitter");
                        Thread.Sleep(Jitter.JitterFactorDelay(StayOpen, JitterFactor));

                        Infrastructure.WorkingHours.Is(handler);

                        // in case just came from a break, wait a little bit before immediately saving the file
                        Thread.Sleep(Jitter.JitterFactorDelay(StayOpen / 4, JitterFactor));

                        var rand = RandomFilename.Generate();

                        var defaultSaveDirectory = timelineEvent.CommandArgs[0].ToString();
                        if (defaultSaveDirectory.Contains("%"))
                        {
                            defaultSaveDirectory = Environment.ExpandEnvironmentVariables(defaultSaveDirectory);
                        }

                        try
                        {
                            foreach (var key in timelineEvent.CommandArgs)
                            {
                                if (key.ToString().StartsWith("save-array:"))
                                {
                                    var savePathString =
                                        key.ToString().Replace("save-array:", "").Replace("'", "\"");
                                    savePathString =
                                        savePathString.Replace("\\",
                                            "/"); //can't seem to deserialize windows path \
                                    var savePaths = JsonConvert.DeserializeObject<string[]>(savePathString);
                                    defaultSaveDirectory =
                                        savePaths.PickRandom().Replace("/", "\\"); //put windows path back
                                    if (defaultSaveDirectory.Contains("%"))
                                    {
                                        defaultSaveDirectory = Environment.ExpandEnvironmentVariables(defaultSaveDirectory);
                                    }
                                    break;
                                }
                            }
                        }
                        catch (Exception e)
                        {
                            Log.Trace($"save-array exception: {e}");
                        }

                        defaultSaveDirectory = ApplicationDetails.GetPath(defaultSaveDirectory);

                        if (!Directory.Exists(defaultSaveDirectory))
                        {
                            Directory.CreateDirectory(defaultSaveDirectory);
                        }

                        var path = $"{defaultSaveDirectory}\\{rand}.docx";

                        //if directory does not exist, create!
                        Log.Trace($"Checking directory at {path}");
                        var f = new FileInfo(path).Directory;
                        if (f == null)
                        {
                            Log.Trace($"Directory does not exist, creating directory at {f.FullName}");
                            Directory.CreateDirectory(f.FullName);
                        }

                        try
                        {
                            if (File.Exists(path))
                            {
                                File.Delete(path);
                            }
                        }
                        catch (Exception e)
                        {
                            Log.Debug(e);
                        }

                        document.Saved = true;
                        if (string.IsNullOrEmpty(document.Path))
                        {
                            document.SaveAs(path);
                            FileListing.Add(path, handler.HandlerType);
                            Log.Trace($"{handler.HandlerType} saving new file: {path}");
                        }
                        else
                        {
                            document.Save();
                            Log.Trace($"{handler.HandlerType} saving existing file: {document.FullName}");
                        }

                        Report(handler.HandlerType.ToString(), timelineEvent.Command,
                            timelineEvent.CommandArgs[0].ToString());
                            

                        if (timelineEvent.CommandArgs.Contains("pdf"))
                        {
                            // randomly generate pdf
                            if (_random.Next(100) < 40)
                            {
                                // Save document into PDF Format
                                object oMissing = System.Reflection.Missing.Value;
                                object outputFileName = timelineEvent.CommandArgs.Contains("pdf-vary-filenames")
                                    ? $"{defaultSaveDirectory}\\{RandomFilename.Generate()}.pdf"
                                    : path.Replace(".docx", ".pdf");
                                object fileFormat = WdSaveFormat.wdFormatPDF;

                                document.SaveAs(outputFileName, fileFormat, oMissing, oMissing,
                                    oMissing, oMissing, oMissing, oMissing,
                                    oMissing, oMissing, oMissing, oMissing,
                                    oMissing, oMissing, oMissing, oMissing);
                                // end save as pdf
                                Report(handler.HandlerType.ToString(), timelineEvent.Command, "pdf");
                                FileListing.Add(outputFileName.ToString(), handler.HandlerType);
                            }
                        }

                        if (_random.Next(100) < 50)
                            document.Close();

                        //if (timelineEvent.DelayAfter > 0)
                        //{
                        //    //sleep and leave the app open
                        //    Log.Trace($"Sleep after for {timelineEvent.DelayAfter}");
                        //    Thread.Sleep(timelineEvent.DelayAfter - writeSleep);
                        //}

                        document.Dispose();
                        document = null;

                        wordApplication.Quit();

                        try
                        {
                            wordApplication.Dispose();
                        }
                        catch
                        {
                            // ignore
                        }
                            
                        try
                        {
                            Marshal.ReleaseComObject(wordApplication);
                        }
                        catch
                        {
                            // ignore
                        }

                        try
                        {
                            Marshal.FinalReleaseComObject(wordApplication);
                        }
                        catch
                        {
                            // ignore
                        }
                    }

                    GC.Collect();
                }
                catch (ThreadAbortException)
                {
                    KillApp();
                    Log.Trace("Word closing...");
                }
                catch (Exception e)
                {
                    Log.Debug(e);
                }
                finally
                {
                    if (timelineEvent.DelayAfter > 0)
                    {
                        //sleep and leave the app open
                        Log.Trace($"Sleep after for {timelineEvent.DelayAfter} with jitter");
                        // Thread.Sleep(timelineEvent.DelayAfter - writeSleep);
                        Thread.Sleep(Jitter.JitterFactorDelay(timelineEvent.DelayAfter, JitterFactor));
                    }
                    else
                    {
                        Thread.Sleep(5000);
                    }
                }
            }
        }
        catch (ThreadAbortException)
        {
            //ignore
        }
        catch (Exception e)
        {
            Log.Error(e);
        }
        finally
        {
            KillApp();
            Log.Trace("Word closing...");
        }
    }

    public void ParseRandomHandlerArgs(TimelineHandler handler)
    {
        if (handler.HandlerArgs.ContainsKey("stay-open"))
        {
            int.TryParse(handler.HandlerArgs["stay-open"].ToString(), out StayOpen);
        }
        if (handler.HandlerArgs.ContainsKey("delay-jitter"))
        {
            JitterFactor = Jitter.JitterFactorParse(handler.HandlerArgs["delay-jitter"].ToString());
        }
    }
    private WdColor GetWdColor(Color color)
    {
        var rgbColor = VB.Information.RGB(color.R, color.G, color.B);
        var wdColor = (WdColor)rgbColor;
        return wdColor;
    }
}
```

**Comment by sei-dupdyke** [comment]
Thanks, I am working to roll this into existing Office app handlers.

**Comment by sei-dupdyke** [comment]
This is available in release 7.0.152 which is out there now.
